File: GraphicsPanel.py
import random
import pygame
from Cell import Cell


class GraphicsPanel:
    ROWS = 10
    COLS = 10
    CELL_SIZE = 30  # This is used to calculate where cells are drawn
    NUM_BOMBS = 10

    def __init__(self):
        self.width = self.COLS * self.CELL_SIZE
        self.height = self.ROWS * self.CELL_SIZE

        # Create a 2D array of Cell objects
        self.cells = [
            [Cell(col * self.CELL_SIZE, row * self.CELL_SIZE) for col in range(self.COLS)]
            for row in range(self.ROWS)
        ]

        self.place_bombs(self.NUM_BOMBS)
        self.calculate_neighbor_bomb_counts()

    def place_bombs(self, bomb_count):
        self.cells[0][0].set_bomb()

    def calculate_neighbor_bomb_counts(self):
        for row in range(self.ROWS):
            for col in range(self.COLS):
                count = self.count_neighbor_bombs(row, col)
                self.cells[row][col].set_neighbor_bomb_count(count)

    def count_neighbor_bombs(self, row, col):
        count = 0
        # Loop over all neighboring cells (including diagonals)
        for i in range(-1, 2):
            for j in range(-1, 2):
                neighbor_row = row + i
                neighbor_col = col + j
                if 0 <= neighbor_row < self.ROWS and 0 <= neighbor_col < self.COLS:
                    if self.cells[neighbor_row][neighbor_col].has_bomb:
                        count += 1
        return count

    def reveal_cell(self, row, col):
        if 0 <= row < self.ROWS and 0 <= col < self.COLS:
            cell = self.cells[row][col]
            if not cell.is_revealed:
                cell.reveal()
                # If this cell has no neighboring bombs and isn’t a bomb, reveal surrounding cells
                if cell.get_neighbor_bomb_count() == 0 and not cell.has_bomb:
                    self.reveal_surrounding_cells(row, col)

    def reveal_surrounding_cells(self, row, col):
        for i in range(-1, 2):
            for j in range(-1, 2):
                # Skip the current cell
                if i != 0 or j != 0:
                    self.reveal_cell(row + i, col + j)

    def draw(self, surface):
        # Draw each cell onto the provided surface
        for row in self.cells:
            for cell in row:
                cell.draw(surface)

    def handle_mouse_event(self, event):
        if event.type == pygame.MOUSEBUTTONDOWN:
            x, y = event.pos
            col = x // self.CELL_SIZE
            row = y // self.CELL_SIZE
            if event.button == 1:  # Left click
                print("Left button clicked")
                self.reveal_cell(row, col)
            elif event.button == 3:  # Right click
                print("Right button clicked")

File: Java/GraphicsPanel.java
import javax.swing.JPanel;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.util.Random;

public class GraphicsPanel extends JPanel implements MouseListener{
    private Cell[][] cells; // 2D array of Cell objects
    private static final int ROWS = 10;
    private static final int COLS = 10;
    private static final int NUM_BOMBS = 10;

    public GraphicsPanel() {
    	setPreferredSize(new Dimension(ROWS * 30, COLS * 30)); 
        this.addMouseListener(this);
        this.cells = new Cell[ROWS][COLS];

        // Initialize the Cell objects
        for (int row = 0; row < ROWS; row++) {
            for (int col = 0; col < COLS; col++) {
                cells[row][col] = new Cell(col * 30, row * 30); // Assuming 30x30 cell size
            }
        }

        // Randomly place bombs and calculate neighbor bomb counts
        placeBombs(); 
        calculateNeighborBombCounts();
    }

    @Override
    protected void paintComponent(Graphics g) {
        // Draw each cell
        for (int row = 0; row < cells.length; row++) {
            for (int col = 0; col < cells[row].length; col++) {
                cells[row][col].draw(g, this);
            }
        }
    }

    public void revealCell(int row, int col) {
        if (row >= 0 && row < ROWS && col >= 0 && col < COLS) {
            Cell cell = cells[row][col];
            if (!cell.isRevealed()) {
                cell.reveal();
                repaint();

                // If the cell has no neighboring bombs, reveal surrounding cells
                if (cell.getNeighborBombCount() == 0 && !cell.hasBomb()) {
                    revealSurroundingCells(row, col);
                }
            }
        }
    }

    private void revealSurroundingCells(int row, int col) {
        for (int i = -1; i <= 1; i++) {
            for (int j = -1; j <= 1; j++) {
                if (i != 0 || j != 0) 	
                	revealCell(row + i, col + j);
            }
        }
    }

    // this function should randomly place NUM_BOMBS on the board.
    // right now it only places one bomb in the top left corner.
    private void placeBombs() {

    }

    private void calculateNeighborBombCounts() {
        for (int row = 0; row < cells.length; row++) {
            for (int col = 0; col < cells[row].length; col++) {
                int count = countNeighborBombs(row, col);
                cells[row][col].setNeighborBombCount(count);
            }
        }
    }

    // this function should count how many bombs surround the cell located at [row][col]
    private int countNeighborBombs(int row, int col) {
        int count = 0;

        return count;
    }

    @Override
    public void mouseClicked(MouseEvent e) {}

    @Override
    public void mousePressed(MouseEvent e) {
	if (e.getButton() == MouseEvent.BUTTON1) {
            System.out.println("Left button clicked");
            revealCell(e.getY()/30, e.getX()/30);
        } else if (e.getButton() == MouseEvent.BUTTON3) {
            System.out.println("Right button clicked");
        }		
    }

    @Override
    public void mouseReleased(MouseEvent e) {}

    @Override
    public void mouseEntered(MouseEvent e) {}

    @Override
    public void mouseExited(MouseEvent e) {}
}
